### PR TITLE
HPCC-20241 Retrieve User Details from LDAP server

### PR DESCRIPTION
**Pull request by wangkx**
In session based ESP, the ISecUser may not contains the User
Details because ESP does not authenticate every request and
the User Details are not filled in. WsAccount should retrieve
User Details directly from LDAP server.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead please review.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 1]
```diff
@@ -2076,6 +2077,27 @@ class CLdapClient : implements ILdapClient, public CInterface
...
+                        {
+                            CLDAPGetValuesLenWrapper valsLen(ld, message, attribute);
+                            if (!m_domainPwdsNeverExpire && !accountPwdNeverExpires && valsLen.hasValues())
+                            {
+                                CDateTime expiry;
+                                char * val = (char*)valsLen.queryCharValue(0);
+                                calcPWExpiry(expiry, (unsigned)strlen(val), val);
+                                ((CLdapSecUser*)&user)->setPasswordExpiration(expiry);
+                            }
+                        }
```
Comment (RussWhitehead) [inline system/security/LdapSecurity/ldapconnection.cpp]: your search filter specifies distinguishedName, but you are not extracting it

**Review by RussWhitehead — CHANGES_REQUESTED** [review]

**Review by RussWhitehead — APPROVED** [review]

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

**Comment by richardkchapman** [comment]
When is this code called? Any performance concerns? @afishbeck can you sanity-check?

**Comment by afishbeck** [comment]
The changes look ok. But I have a question for @RussWhitehead 

So getUserInfo now always seems to retrieve an extra 3 fields from LDAP.

The particular usage for this JIRA is only to display user info on demand, but

@RussWhitehead per Richards question, how often / where is getUserInfo and retrieveUserInfo (in IPermissionsProcessor) called?  And how much overhead does a few extra fields being retrieved add?

**Comment by RussWhitehead** [comment]
@afishbeck These is called whenever a user logs in, and when they view/change their user info, and when exporting user account information. Since we did not introduce any additional calls to LDAP, there should be no performance issues. 
@richardkchapman Looks good for merge

**Pull request merged**